### DISCUSSION TITLE
Question: how to debug zeroconf when spotify clients don't see librespot?

### DISCUSSION THREAD

**Discussion by Pierric82** [librespot-org/librespot#1314] [General]
Hi,

Until recently I was using raspotify, with credentials since I had never managed (or maybe a long time ago? not sure) to make zeroconf work out of the box. As we know, recently Spotify made credentials-based access impossible. I still didn't manage to make raspotify work so I moved to the source and compiled librespot.

Where I'm at now: whether I include the with-dns-sd feature or not, I can see in the output that a zeroconf server is started. But my device is not visible from Spotify clients on the same local network.

Here's the command line I use and the output:
```
pierric@pi4 ~/s/librespot (dev)> sudo target/release/librespot --device hdmi:CARD=vc4hdmi0,DEV=0 --disable-audio-cache -n "librespot" --verbose -i 0.0.0.0 -c tmpcache/
[2024-08-23T09:59:54Z INFO  librespot] librespot 0.5.0-dev a3a1c17 (Built on 2024-08-23, Build ID: 4uv7lpzv, Profile: release)
[2024-08-23T09:59:54Z TRACE librespot] Command line argument(s):
[2024-08-23T09:59:54Z TRACE librespot]          device "hdmi:CARD=vc4hdmi0,DEV=0"
[2024-08-23T09:59:54Z TRACE librespot]          disable-audio-cache
[2024-08-23T09:59:54Z TRACE librespot]          n "librespot"
[2024-08-23T09:59:54Z TRACE librespot]          verbose
[2024-08-23T09:59:54Z TRACE librespot]          i "0.0.0.0"
[2024-08-23T09:59:54Z TRACE librespot]          c "tmpcache/"
[2024-08-23T09:59:54Z DEBUG librespot_core::session] new Session
[2024-08-23T09:59:54Z DEBUG librespot_discovery::server] Zeroconf server listening on 0.0.0.0:42849
*** WARNING *** The program 'librespot' uses the Apple Bonjour compatibility layer of Avahi.
*** WARNING *** Please fix your application to use the native API of Avahi!
*** WARNING *** For more information see <http://0pointer.de/blog/projects/avahi-compat.html>
[2024-08-23T09:59:54Z INFO  librespot_playback::mixer::softmixer] Mixing with softvol and volume control: Log(60.0)
[2024-08-23T09:59:54Z DEBUG librespot_playback::player] new Player [0]
[2024-08-23T09:59:54Z INFO  librespot_playback::convert] Converting with ditherer: tpdf
[2024-08-23T09:59:54Z INFO  librespot_playback::audio_backend::rodio] Using Rodio sink with format S16 and cpal host: ALSA
ALSA lib pcm_dmix.c:972:(snd_pcm_dmix_open) The dmix plugin supports only playback stream
ALSA lib pcm_asym.c:105:(_snd_pcm_asym_open) capture slave is not defined
ALSA lib pcm_asym.c:105:(_snd_pcm_asym_open) capture slave is not defined
[2024-08-23T09:59:54Z INFO  librespot_playback::audio_backend::rodio] Using audio device: hdmi:CARD=vc4hdmi0,DEV=0
[2024-08-23T09:59:54Z DEBUG librespot_playback::audio_backend::rodio] Rodio sink was created
```

Some notes:
- This particular output is with with-dns-sd enabled at compile time. From memory the only difference if that feature is not included is that the warning messages about Avahi are not displayed.
- I have tried running and stopping the Avahi daemon. In the case of with-dns-sd on and avahi-daemon stopped, I get the following error:
```
[2024-08-23T10:25:35Z DEBUG librespot] System uptime > 1 min, not retrying to initialise discovery
[2024-08-23T10:25:35Z WARN  librespot] Could not initialise discovery: Service unavailable { DNS-SD Error: Unknown }
[2024-08-23T10:25:35Z ERROR librespot] Discovery is unavailable and no credentials provided. Authentication is not possible.
```
however in with dns-sd off, avahi being on or off has no impact
- The device this runs on is a Raspberry Pi 4, running Raspberry Pi OS 12.
- It is connected to more than one local network, the main one via Ethernet, but I believe that with the -i 0.0.0.0 option that should not be a problem: it is connected to the 192.168.1.0 network that my wifi devices connect to as well. Am I wrong with that assumption?

I have never understood how zeroconf, avahi, dns-sd, etc. *really* work, especially in the context of a Spotify connect device. I would love to get a good explanation on this if it exists anywhere. Either way, would someone be able to tell me how to further debug the situation? are there ways to emulate spotify searching for devices and see where it goes wrong? I'm stuck and don't know how to further look into this.

Thanks!
Pierric.

**Comment by Pierric82** [comment]
Some extra information: I installed avahi-utils and was able to browse for Spotify connect (i.e. librespot):
```
pierric@pi4 ~/s/librespot (dev)> avahi-browse -r _spotify-connect._tcp
+   eth0 IPv4 librespot                                     _spotify-connect._tcp local
=   eth0 IPv4 librespot                                     _spotify-connect._tcp local
   hostname = [pi4.local]
   address = [192.168.1.32]
   port = [40285]
   txt = ["CPath=/" "VERSION=1.0"]
```

It also works from other linux servers in the LAN (all on Ethernet). But my Spotify clients (Wifi, but same LAN) still don't see anything. Initially pi4.local would not resolve correctly so I set up PiHole to have a fixed entry with the corresponding entry for that. `ping pi4.local` now works.

My MacOS computer on the WiFi _does_ seem to see the client, too:
```
❯ dns-sd -B _spotify-connect._tcp
Browsing for _spotify-connect._tcp
DATE: ---Fri 23 Aug 2024---
16:42:24.698  ...STARTING...
Timestamp     A/R    Flags  if Domain               Service Type         Instance Name
16:42:24.873  Add        2  12 local.               _spotify-connect._tcp. librespot
```

Still, Spotify clients don't see it.

Based on what I read [here](https://developer.spotify.com/documentation/commercial-hardware/implementation/guides/zeroconf), I tried this command to see what would happen:
```
pierric@pi4 ~/s/librespot (dev)> curl -G "http://pi4.local:40285/cpath?action=getInfo&version=2.10.0"
{"accountReq":"PREMIUM","activeUser":"","availability":"","brandDisplayName":"librespot","clientID":"65b708073fc0480ea92a077233ca87bd","deviceID":"7c28ab8a5c9512e4266ac7cb756312c82ee43d7e","deviceType":"Speaker","groupStatus":"NONE","libraryVersion":"0.5.0-dev","modelDisplayName":"librespot","productID":0,"publicKey":"WhJUXPqtLIggBGH0Fo08kSF20mc9OKcByZ+MAh6fkGZxLB8wrGEnl8HMgw9V1M2bdngNN4rL52DqmlM6snN6ygyi85epBxza/aO68bmgbscV0p02Lld/KJEf9wllzfpv","remoteName":"librespot","resolverVersion":"1","scope":"streaming","spotifyError":0,"status":101,"statusString":"OK","supported_capabilities":1,"supported_drm_media_formats":[],"tokenType":"default","version":"2.9.0"}
```

I can also open that URL successfully in a browser on my mobile phone whose Spotify app cannot see the device. But I'm not sure what to do next.

EDIT: Well I'm speechless. I left everything hanging as it was after spending a few hours getting to the situation described above, and went for an errand. I came back home, opened spotify, and librespot was visible .... I would still like to understand better what was failing and is now working, to ensure it *keeps* working...

**Comment by polyrainbow** [comment]
Same issue here: librespot's mDNS advertising (checked with `avahi-browse --all` from another device), and HTTP server work in the local network, but Spotify clients do not show librespot. I have kept the service running for several hours, but no difference. Using librespot v0.5.0

**Reply by Pierric82** [reply]
Does it show as
`+   eth0 IPv4 librespot                                     _spotify-connect._tcp local`
in the output? That's how I see it here now that it's working.

**Reply by polyrainbow** [reply]
Output is
```
+   wlo1 IPv6 Librespot                                     _spotify-connect._tcp local
+   wlo1 IPv4 Librespot                                     _spotify-connect._tcp local
```
(client is connected via WiFi interface)

**Reply by Pierric82** [reply]
Playing around with this on my side to find out how to help more, what do you see with
`avahi-browse  -vr _spotify-connect._tcp`
?

On my side it gives:
```
+   eth0 IPv4 raspotify (pi4)                               _spotify-connect._tcp local
=   eth0 IPv4 raspotify (pi4)                               _spotify-connect._tcp local
   hostname = [pi4.local]
   address = [192.168.1.240]
   port = [44265]
   txt = ["CPath=/" "VERSION=1.0"]
```

(never mind the mention of raspotify instead of librespot, I've been switching back and forth)

Considering the parameters above I can then simulate that I'm the spotify app enquiring about a device by calling 
`curl -G "http://192.168.1.240:44265/cpath?action=getInfo&version=2.10.0"`
which returns some json with information about the (virtual) device.

Try and see if you can do the same, it may point you towards what's failing?

**Reply by polyrainbow** [reply]
After digging around for a few more hours, I think I have found the issue: I'm using a Mesh WiFi system (TP-Link Deco) which seems to not reliably forward mDNS packages. Everything works only if the client is in the same cell as the librespot server.
The above `avahi-browse` output might have been recorded shortly before roaming from the librespot cell to another one. [A person with a similar issue was claiming that it works when switching the Mesh system from router mode to AP mode.](https://www.reddit.com/r/homeassistant/comments/1cj1fbj/mesh_wifi_recommendations/) So I tried the same, set up an OpenWRT router on a Raspberry Pi, and now it works.

Thanks for your help @Pierric82 , very much appreciated.

**Reply by Pierric82** [reply]
Glad you got to the bottom of it!
